jax/experimental/jax2tf/README.md

```diff
@@ -1005,12 +1005,27 @@ The `jax2tf`-lowered function supports higher-order gradients, but when the
 function is saved in a SavedModel, only the first-order gradient is saved.
 This is primarily a limitation of the SavedModel support for custom gradients.
 
+### Native serialization supports only select dialects
+
+Applies to native serialization only.
+
+JAX native serialization checks that the code to be serialized contains
+operations only from MLIR dialects that are known to have stability guarantees,
+e.g., StableHLO.
+
+If your code uses custom JAX lowering rules that use operations from other
+dialects, and if you have reasons to believe that those operations are stable
+over the planned lifetime of the serialized artifact,
+you can disable the safety check by adding
+`jax2tf.DisabledSafetyCheck.allowed_dialect("my_dialect")` to the `disabled_checks`
```

Comment:
Moving this discussion here:

> > > I am wondering how important is this safety mechanism. Perhaps it is very hard or impossible for serialize_portable_artifact to let through non-StableHLO (and non-MHLO) dialect ops. Can we count on that? I In that case, the thing that I can do instead is to ensure that the custom_call check works for mhlo.custom_call also?
> > 
> > 
> > serializePortableArtifact _only_ lets through valid programs. You are correct about being able to register dialects, but [`createStablehloLegalizeToVhloPass`](https://github.com/openxla/stablehlo/blob/3165471a4b9026e6912d4dd957ae8c02152a3b86/stablehlo/dialect/Serialization.cpp#L41) will fail if any part of the program is not guaranteed to be stable, including other dialects, unsupported ops and types.
> > Also as far as filtering by dialect name goes - we support a subset of both mhlo and builtin dialect in StableHLO. For example, if a program somehow has a `memref` in it, serializePortableArtifact will fail. That said, you will catch the vast majority of issues by filtering on dialect name before calling serializePortableArtifact.
> 
> It seems then that you would suggest to NOT add these checks, and rely instead of serializePortableArtifact?

I am suggesting it is _safe_ to NOT add these checks, I could go either way on re-validating at the JAX level. The benefit of adding them at the JAX layer would be that you have finer-grained control over the error messages. We do have an action item to improve our error reporting in https://github.com/openxla/stablehlo/issues/680.

On a separate note, I'm not sure this `allowed_dialect` filtering will work for `"my_dialect"`, since StableHLO portable artifacts do not support arbitrary dialects - serializePortableArtifact will fail. This works for MHLO/CHLO since we convert these dialects to StableHLO before making the portable artifact in `xla_extension`. If stability is not required for these mixed-dialect cases, another option would be just to use regular MLIR bytecode, instead of StableHLO portable artifact (which is MLIR Bytecode + compatibility guarantees).

Reply:
If the checks are not necessary, and if the `allow_dialect` is not an effective escape hatch, I will abandon this change.

I will propose a smaller change instead that plugs the loophole I had found, where some code uses `mhlo.custom_call` for an unsupported custom call and this is not detected by the current code.

Reply:
That sounds good! Off the top of my head, I don't think there are other loopholes aside from custom call of "things that are valid in the MHLO->StableHLO converter that we don't want to permit." So targeted validation should be enough.

Reply:
I have redone this, it is much simpler. @GleasonK PTAL